lib/client-request.js

```diff
@@ -0,0 +1,176 @@
+    }
+
+    h += '\r\n'
+
+    return h
+  }
+
+  _final (cb) {
+    if (this.headersSent === false) this.flushHeaders()
+
+    this._pendingFinal = cb
+  }
+
+  _ondata (data) {
+    if (!this._responseHeadComplete) {
+      const blankLine = Buffer.from([13, 10, 13, 10])
+      const blankLineIndex = data.indexOf(blankLine)
+      const hasBlankLine = blankLineIndex !== -1
+
+      const h = hasBlankLine ? data.subarray(0, blankLineIndex) : data
+
+      if (this._responseHeadBuffer !== null) {
+        this._responseHeadBuffer = Buffer.concat([this._responseHeadBuffer, h])
+      } else {
+        this._responseHeadBuffer = h
+      }
+
+      if (hasBlankLine) {
+        this._responseHeadComplete = true
+        this._readHead(this._responseHeadBuffer)
+        data = data.subarray(blankLineIndex + blankLine.byteLength)
+      }
+    }
+
+    if (this._responseHeadComplete) this._readBody(data)
+  }
+
+  _readHead (head) {
+    const h = head.toString().split('\r\n')
+
+    let [, statusCode, statusMessage] = h[0].split(' ')
+    if (!statusCode || !statusMessage) return this._socket.destroy()
+
+    statusCode = +statusCode
+
+    const headers = {}
+
+    for (let i = 1; i < h.length; i++) {
+      const [name, value] = h[i].split(': ')
+      headers[name.toLowerCase()] = value
+    }
+
+    this._response = new HTTPIncomingMessage(this._socket, headers, { statusCode, statusMessage })
+
+    this.emit('response', this._response)
+  }
+
+  _readBody (data) {
+    const len = Number(this._response.getHeader('content-length')) || 0
+    const chunked = this._response.getHeader('transfer-encoding') === 'chunked'
+    const chunkedTail = Buffer.from([48, 13, 10, 13, 10]) // '0\r\n\r\n'
+
+    if (this._responseBodyBuffer !== null) {
+      this._responseBodyBuffer = Buffer.concat([this._responseBodyBuffer, data])
+    } else {
+      this._responseBodyBuffer = data
+    }
+
+    const hasChunkedTail = chunked && this._responseBodyBuffer.indexOf(chunkedTail) !== -1
+    if (hasChunkedTail) {
+      const slices = []
+      let sliceStart = -1
+      let readingSlice = false
+
+      for (let i = 0; i < this._responseBodyBuffer.byteLength - 5; i++) {
+        if (readingSlice && sliceStart === -1) {
+          sliceStart = i
+        } else if (this._responseBodyBuffer[i] === 10 && this._responseBodyBuffer[i - 1] === 13) {
+          if (readingSlice && sliceStart !== -1) {
+            slices.push([sliceStart, i - 1])
+            sliceStart = -1
+          }
+
+          readingSlice = !readingSlice
+        }
+      }
+
+      const messageSlices = slices.map(([s, e]) => this._responseBodyBuffer.subarray(s, e))
+      this._responseBodyBuffer = Buffer.concat(messageSlices)
+    }
+
+    if ((!chunked && this._responseBodyBuffer.byteLength === len) || hasChunkedTail) {
+      if (this._responseBodyBuffer.byteLength > 0) this._response.push(this._responseBodyBuffer)
+      this._response.push(null)
+      this._socket.destroy()
+    }
```

Comment:
This will read the body in its entirety before pushing it to the stream, which isn't ideal as it will buffer the entire body in memory before passing it along. Instead, we should push chunks as they arrive, buffering only the bare minimum between TCP fragments.

Reply:
That's true, I pushed a new commit and the current behavior is:
- For requests with `content-length`: Any data available is sent immediately.
- For chunked messages: Chunks are sent when they are complete, individually, that is, big chunks will be kept in memory similarly as before, the unit test for receiving big writes has only 3 chunks, for instance.

I am trying a new implementation to send "chunks in chunks" but the state management + parsing lines + edge cases handling is getting a bit wild.